qadence/ml_tools/callbacks/writer_registry.py

```diff
@@ -0,0 +1,375 @@
+
+        return self.mlflow
+
+    def close(self) -> None:
+        """Closes the MLflow run."""
+        if self.run:
+            self.mlflow.end_run()
+
+    def write(self, result: OptimizeResult) -> None:
+        """
+        Logs the results of the current iteration to MLflow.
+
+        Args:
+            result (OptimizeResult): The optimization results to log.
+        """
+        # Not writing loss as loss is available in the metrics
+        # if result.loss is not None:
+        #     self.mlflow.log_metric("loss", float(result.loss), step=result.iteration)
+        if self.mlflow:
+            self.mlflow.log_metrics(result.metrics, step=result.iteration)
+
+    def log_hyperparams(self, hyperparams: dict) -> None:
+        """
+        Logs hyperparameters to MLflow.
+
+        Args:
+            hyperparams (dict): A dictionary of hyperparameters to log.
+        """
+        if self.mlflow:
+            self.mlflow.log_params(hyperparams)
+
+    def plot(
+        self,
+        model: Module,
+        iteration: int,
+        plotting_functions: tuple[PlottingFunction, ...],
+    ) -> None:
+        """
+        Logs plots of the model using provided plotting functions.
+
+        Args:
+            model (Module): The model to plot.
+            iteration (int): The current iteration number.
+            plotting_functions (tuple[PlottingFunction, ...]): Functions used
+                to generate plots.
+        """
+        if self.mlflow:
+            for pf in plotting_functions:
+                descr, fig = pf(model, iteration)
+                self.mlflow.log_figure(fig, descr)
+
+    def log_model(
+        self,
+        model: Module,
+        dataloader: Union[None, DataLoader, DictDataLoader],
+    ) -> None:
+        """
+        Logs the model and its signature to MLflow.
+
+        Args:
+            model (Module): The model to log.
+            dataloader (DataLoader | DictDataLoader | None): DataLoader to
+                use for model input.
+        """
+        if self.mlflow:
+            signature = None
+            if dataloader is not None:
+                xs: InputData
+                xs, *_ = next(iter(dataloader))
+                preds = model(xs)
+                if isinstance(xs, Tensor):
+                    xs = xs.detach().cpu().numpy()
+                    preds = preds.detach().cpu().numpy()
+                elif isinstance(xs, dict):
+                    xs = {key: val.detach().cpu().numpy() for key, val in xs.items()}
+                    preds = {key: val.detach().cpu().numpy() for key, val in preds.items()}
+                try:
+                    from mlflow.models import infer_signature
+
+                    signature = infer_signature(xs, preds)
+                except ImportError:
+                    logger.warning(
+                        "MLflow's infer_signature is not available. Please install mlflow."
+                    )
```

Comment:
This looks a bit long and nested. Can we split it?

Reply:
Updated.